### PR TITLE
Feature/constraints

### PR DESCRIPTION
**Pull request by tbreeds**
Add native constraints support

Constraints are a new feature in pip 9.x, designed to enforce a specific ''constrainted'' set of installed libraries.  This option when paired with ''{constratints}'' in your install_command will instruct pip where to get the constratints from.

It's possible to use constraints without this change by manually adding:
    -c /path/to/constraints/file.txt
into install_coommand.  However this means that the constraints are used when doing the develop-inst.  This is fine *unless* the project you're trying to test is listed in your constraints file.  Then pip errors out saying that you can't satisfy the constraint from the dev install.

So we only use constraints in the installdeps phase

### CODE REVIEW

**Comment by obestwalter** [comment]
He @tbreeds, thank you for your PR. 

@loechel as you added the docs about constraints recently and might want to have a look?

**Comment by tbreeds** [comment]
@obestwalter Do I need to address the travis failure?  It looks unrelated to my patch

**Comment by obestwalter** [comment]
I don't see a failure.

**Comment by loechel** [comment]
@tbreeds / @obestwalter first, I have restarted the test, and it is now passing, so that is nothing to be handled.

I will have a more detailed look this weekend, and will make my comments and a detailed code review.

**Comment by obestwalter** [comment]
Hi @loechel - nice - thanks :)

**Comment by tbreeds** [comment]
Thanks @loechel

File: doc/config.rst [hunk 1]
```diff
@@ -89,6 +89,15 @@ Complete list of settings that you can put into ``testenv*`` sections:
     Similar to ``make`` recipe lines, any command with a leading ``-``
     will ignore the exit code.
 
+.. confval:: constraints=STRING
+
+    .. versionadded:: 2.10
+
+    Constraints are a new feature in pip 9.x, designed to enforce a specific
```
Comment (loechel) [inline doc/config.rst]: Constraints where added by pip 7.1 --> https://pip.pypa.io/en/stable/user_guide/#constraints-files

File: tests/test_z_cmdline.py [hunk 2]
```diff
@@ -643,10 +643,12 @@ def test_hello(pytestconfig):
 
 def _alwayscopy_not_supported():
     # This is due to virtualenv bugs with alwayscopy in some platforms
-    # see: https://github.com/pypa/virtualenv/issues/565
+    # See: https://github.com/pypa/virtualenv/issues/565
```
Comment (loechel) [inline tests/test_z_cmdline.py]: This fix is unrelated to this feature, should probably go into another bugfix pull request

File: tox/venv.py [hunk 3]
```diff
@@ -289,6 +290,14 @@ def run_install_command(self, packages, action, options=()):
             i = argv.index('{opts}')
             argv[i:i + 1] = list(options)
 
+        if '{constraints}' in argv:
```
Comment (loechel) [inline tox/venv.py]: If I see it correct, this is the actual feature improvement implementation part. 
It seems to be correct, that it adds ``-c constraints`` to the install_command list.

But I did not see anything about the described cases, that the tested package itself is listed in the constraints file. Could you please correct me, but I did not see anything that makes a different to add a file to the constraints parameter than adding -cconstraints.txt to deps parameter. 

It might be more declarative and explicit that before, but I did not see the described case handled.

Comment (tbreeds) [reply tox/venv.py]: I think you're correct.  Adding -c to the deps does seem to do handle the situation I'm trying to deal with.

Should I remove this PR while I investigate?

File: tests/test_z_cmdline.py (hunk 2 above)
Comment (tbreeds) [reply tests/test_z_cmdline.py]: Okay I'll do that.  I thought that having it in a separate commit was adequate.

**Comment by tbreeds** [comment]
@asottile install_command and setenv impact all tox phases and leads something like:

```
[tony@thor tox]$ cat constraints.txt ; git diff tox.ini
apipkg===1.4
coverage===4.4.1
execnet===1.4.1
pluggy===0.5.2
py===1.4.34
pytest===3.2.2
pytest-cov===2.5.1
pytest-forked===0.2
pytest-timeout===1.2.0
pytest-xdist===1.20.0
six===1.11.0
tox===2.8.0
virtualenv===15.1.0
diff --git a/tox.ini b/tox.ini
index 303bd82..8922a84 100644
--- a/tox.ini
+++ b/tox.ini
@@ -13,6 +13,7 @@ deps = pytest >= 3.0.0
 ;NOTE --cov-config={toxinidir}/tox.ini is necessary until
 ;https://github.com/pytest-dev/pytest-cov/issues/168 is fixed
 commands = pytest {posargs:tests} --cov-config={toxinidir}/tox.ini --cov={envsitepackagesdir}/tox --timeout=180
+install_command = pip install -c{toxinidir}/constraints.txt {opts} {packages}
 
 [testenv:dev]
 description = DEV environment, if no posarg is specified: run pytest
```

```
[tony@thor tox]$ tox -re py27 --notestGLOB sdist-make: /home/tony/projects/openstack/related-projects/tox/setup.py
py27 recreate: /home/tony/projects/openstack/related-projects/tox/.tox/py27
py27 installdeps: pytest >= 3.0.0, pytest-cov, pytest-timeout, pytest-xdist
py27 inst: /home/tony/projects/openstack/related-projects/tox/.tox/dist/tox-2.9.0rc2.dev4+g15ec7c4.d20170928.zip
ERROR: invocation failed (exit code 1), logfile: /home/tony/projects/openstack/related-projects/tox/.tox/py27/log/py27-2.log
ERROR: actionid: py27
msg: installpkg
cmdargs: ['/home/tony/projects/openstack/related-projects/tox/.tox/py27/bin/pip', 'install', '-c/home/tony/projects/openstack/related-projects/tox/constraints.txt', '/home/tony/projects/openstack/related-projects/tox/.tox/dist/tox-2.9.0rc2.dev4+g15ec7c4.d20170928.zip']
<SNIP>
Processing ./.tox/dist/tox-2.9.0rc2.dev4+g15ec7c4.d20170928.zip
Could not satisfy constraints for 'tox': installation from path or url cannot be constrained to a version

py27 installed: apipkg==1.4,coverage==4.4.1,execnet==1.4.1,py==1.4.34,pytest==3.2.2,pytest-cov==2.5.1,pytest-forked==0.2,pytest-timeout==1.2.0,pytest-xdist==1.20.0
___________________________________ summary ____________________________________
  py27: skipped tests
  congratulations :)
```
But adding the constraints to the deps section does seem to work.

**Comment by asottile** [comment]
@tbreeds yeah my leaning is toward just doing it through `deps` (plus it's already documented via #585)
